rfcs/20180920-unify-rnn-interface.md

```diff
@@ -0,0 +1,377 @@
+# Unify RNN Inteface
+
+| Status        | Proposed                                             |
+| ------------- |:---------------------------------------------------- |
+| **Author(s)** | @qlzh727 (scottzhu@google.com)                       |
+| **Sponsor**   | @ebrevdo (ebrevdo@google.com), @drpng (drpng@google.com)|
+| **Updated**   | 2018-08-24                                           |
+
+## Objective
+
+Unify RNN (LSTM/GRU/other recurrent models) interfaces between TF RNN and Keras RNN at the release 
+of TensorFlow 2.0.
+
+## Motivation
+
+Recurrent neural networks [RNN](https://en.wikipedia.org/wiki/Recurrent_neural_network) is the deep 
+neural network that works great with sequential input data like text and voice. There are two sets 
+of APIs in TensorFlow that allows users to build a RNN. The APIs are a bit different from each 
+other, but trying to solve the same problem. This situation creates some user confusions about:
+
+1. What's the recommended API to use.
+1. What's the relationship between those 2 APIs and how they work together (which they don't 
+currently).
+
+Ideally those two APIs should be unified into "one true API", which should be:
+1. Intuitive to use (hide unnecessary complexity from user).
+1. Covers most of the common use cases.
+1. Easy to migrate from either side.
+
+Keras has become the recommended high level API for Tensorflow since its easy to understand and use.
+We would prefer to unify the final API that is similar to existing Keras API, and port 
+functionalities from TF RNN to Keras.
+
+
+## Design Proposal
+
+1. Update RNN cells in Keras and TF RNN, so that all of them will work with Keras RNN layer and TF 
+RNN layer.
+1. Remove the duplicated RNN cells between Keras and TF, update the documentation for the migration 
+path.
+    1. tf.nn.rnn_cell.BasicRNNCell -> tf.keras.SimpleRNNCell
+    1. tf.nn.rnn_cell.BasicLSTMCell,  tf.keras.LSTMCell -> tf.nn.rnn_cell.LSTMCell
```

Comment:
If keras becomes the prefered API, why the remaining LSTM cell lives in tf.nn?

Reply:
Thanks for pointing this out. The reason is that tf.nn.LSTMCell supports more features like state clipping and peephole. We might eventually update the tf.keras.LSTMCell to support those features, so that all the keras cells will be preferred.

Reply:
ok thanks for the clarification.